testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java

```diff
@@ -481,19 +551,96 @@ public void testReferencedTablesWithRowFilter()
         assertThat(tables).hasSize(2);
 
         TableInfo table = tables.get(0);
-        assertTableInfoBasics(table, "tpch", "tiny", "nation", "user", false, 1, 0);
+        assertTableInfoBasics(table, "tpch", "tiny", "nation", "user", false, 1, 0, 1);
+
+        TableReferenceInfo filterInfo = Iterables.getOnlyElement(table.getReferenceChain());
+        assertFilterMaskInfo(filterInfo, ROW_FILTER, "(EXISTS (SELECT 1 FROM nation WHERE (name = test_varchar)))", "mock", "default", "test_table_with_row_filter", Optional.empty());
+
+        ColumnInfo column = Iterables.getOnlyElement(table.getColumns());
+        assertThat(column.getColumn()).isEqualTo("name");
+        assertThat(column.getMask()).isEmpty();
+
+        table = tables.get(1);
+        assertTableInfoBasics(table, "mock", "default", "test_table_with_row_filter", "user", true, 1, 1, 0);
+        assertThat(Iterables.getOnlyElement(table.getFilters())).isEqualToIgnoringWhitespace("(EXISTS (SELECT 1 FROM nation WHERE (name = test_varchar)))");
+
+        column = Iterables.getOnlyElement(table.getColumns());
+        assertThat(column.getColumn()).isEqualTo("test_varchar");
+        assertThat(column.getMask()).isEmpty();
+    }
+
+    @Test
+    public void testReferencedTablesWithNestedView()
+            throws Exception
+    {
+        QueryEvents queryEvents = runQueryAndWaitForEvents("SELECT 1 FROM mock.default.test_view_nesting").getQueryEvents();
+
+        QueryCompletedEvent event = queryEvents.getQueryCompletedEvent();
+
+        List<TableInfo> tables = event.getMetadata().getTables();
+        assertThat(tables).hasSize(3);
+
+        TableInfo table = tables.get(0);
+        assertTableInfoBasics(table, "tpch", "tiny", "nation", "user", false, 1, 0, 2);
+
+        TableReferenceInfo viewInfo = table.getReferenceChain().get(0);
+        assertViewInfo(viewInfo, "mock", "default", "test_view", false);
+
+        viewInfo = table.getReferenceChain().get(1);
+        assertViewInfo(viewInfo, "mock", "default", "test_view_nesting", false);
+
+        ColumnInfo column = Iterables.getOnlyElement(table.getColumns());
+        assertThat(column.getColumn()).isEqualTo("nationkey");
+        assertThat(column.getMask()).isEmpty();
+
+        table = tables.get(1);
+        assertTableInfoBasics(table, "mock", "default", "test_view", "user", false, 1, 0, 1);
+        assertViewInfo(Iterables.getOnlyElement(table.getReferenceChain()), "mock", "default", "test_view_nesting", false);
+
+        column = Iterables.getOnlyElement(table.getColumns());
+        assertThat(column.getColumn()).isEqualTo("test_column");
+        assertThat(column.getMask()).isEmpty();
+
+        table = tables.get(2);
+        assertTableInfoBasics(table, "mock", "default", "test_view_nesting", "user", true, 0, 0, 0);
+    }
+
+    @Test
+    public void testReferencedTablesWithRowFilterAndView()
```

Comment:
Trino has a concept of redirection in case of tables - Can we add a test to validate the behaviour in case of redirected tables ?

Reply:
Yup I'm very familiar with redirects (more than I want to be, if I'm being honest...). Added a test for that.

Reply:
Yup I'm very familiar with redirects (more than I want to be, if I'm being honest...). Added a test for that.